shared/extd_nw_types.py

```diff
@@ -3,21 +3,257 @@
+
+        if isinstance(ip_header, IPv4):
+            source_list = [int(octet) for octet in ip_header.source.split(".")]
+            destination_list = [int(octet) for octet in
+                                ip_header.destination.split(".")]
+            source_upper = (source_list[0] << 8) + source_list[1]
+            source_lower = (source_list[2] << 8) + source_list[3]
+            destination_upper = (destination_list[0] << 8) + destination_list[1]
+            destination_lower = (destination_list[2] << 8) + destination_list[3]
+            ip_value = (source_upper + source_lower + destination_upper +
+                        destination_lower + ip_header.protocol + (self.h_length * 4))
+        else:
+            # TODO: Implement IPv6 header analysis
+            ip_value = 0
+
+        block_sum = (self.source_p + self.destination_p + sequence_upper +
+                         sequence_lower + acknowledgment_upper + acknowledgment_lower +
+                         self._length_flags + self.window + 0 + self.urgent_pointer +
+                         self._value_by_half_word(self.options) +
+                         self._value_by_half_word(self.data) + ip_value)
+        
+        while block_sum > 65535:
+            carry = block_sum >> 16
+            block_sum = (block_sum & 65535) + carry
+
+        self.checksum = ~block_sum & 65535
+
+    def pack(self, ip_header: Union[IPv4, IPv6], value=None):
+        """Pack the struct in a binary representation.
+
+        Merge some fields to ensure correct packing.
+
+        Returns:
+            bytes: Binary representation of this instance.
+
+        """
+        self._update_checksum(ip_header)
+        return super().pack(value)
+    
+    def unpack(self, buff:bytes, offset=0):
+        """Unpack a binary struct into this object's attributes.
+
+        Return the values instead of the lib's basic types.
+
+        Args:
+            buff (bytes): Binary buffer.
+            offset (int): Where to begin unpacking.
+
+        Raises:
+            :exc:`~.exceptions.UnpackException`: If unpack fails.
+
+        """
+        super().unpack(buff, offset)
+        self.source_p = self.source_p.value
+        self.destination_p = self.destination_p.value
+        self.sequence_n = self.sequence_n.value
+        self.acknowledgment_n = self.acknowledgment_n.value
+        self.h_length = self._length_flags.value >> 12
+        self.flags = self._length_flags.value & 4095
+        self.window = self.window.value
+        self.checksum = self.checksum.value
+
+        if self.h_length > 5:
+            option_size = (self.h_length - 5) * 4
+            self.data = self.options.value[option_size:]
+            self.options = self.options.value[:option_size]
+        else:
+            self.data = self.options.value
+            self.options = b''
 
 
 class UDP(GenericStruct):
-    """ UDP - To be developed """
+    """ UDP - Being developed """
```

Comment:
Let's include the ascii representation https://www.ietf.org/rfc/rfc768.txt in the docstring 


```

                                    
                  0      7 8     15 16    23 24    31  
                 +--------+--------+--------+--------+ 
                 |     Source      |   Destination   | 
                 |      Port       |      Port       | 
                 +--------+--------+--------+--------+ 
                 |                 |                 | 
                 |     Length      |    Checksum     | 
                 +--------+--------+--------+--------+ 
                 |                                     
                 |          data octets ...            
                 +---------------- ...                 

                      User Datagram Header Format
```

Reply:
Added